File: common/src/main/java/io/github/xiewuzhiying/vs_addition/compats/computercraft/peripherals/CannonMountPeripheral.java
package io.github.xiewuzhiying.vs_addition.compats.computercraft.peripherals;

import dan200.computercraft.api.lua.IArguments;
import dan200.computercraft.api.lua.LuaException;
import dan200.computercraft.api.lua.LuaFunction;
import dan200.computercraft.api.peripheral.IPeripheral;
import io.github.xiewuzhiying.vs_addition.mixin.createbigcannons.CannonMountBlockEntityAccessor;
import net.minecraft.core.BlockPos;
import net.minecraft.world.level.Level;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import rbasamoyai.createbigcannons.cannon_control.cannon_mount.CannonMountBlockEntity;

public class CannonMountPeripheral implements IPeripheral {

    public final String type ;
    public final CannonMountBlockEntity tileEntity;

    public final Level level;

    public final BlockPos worldPosition;

    public CannonMountPeripheral(String type, CannonMountBlockEntity tileEntity, Level level, BlockPos blockPos) {
        this.type = type;
        this.tileEntity = tileEntity;
        this.level = level;
        this.worldPosition = blockPos;
    }

    @NotNull
    @Override
    public String getType() {
        return type;
    }

    @Override
    public boolean equals(@Nullable IPeripheral iPeripheral) {
        return iPeripheral == this;
    }

    @Override
    public Object getTarget() {
        return this.tileEntity;
    }

    @LuaFunction(mainThread = true)
    public final Object assemble(){
        if(!this.tileEntity.isRunning()) {
            ((CannonMountBlockEntityAccessor) this.tileEntity).Assemble();
            return true;
        }
        return false;
    }

    @LuaFunction(mainThread = true)
    public final Object disassemble() {
        if(this.tileEntity.isRunning()) {
            this.tileEntity.disassemble();
            this.tileEntity.sendData();
            return true;
        }
        return false;
    }

    @LuaFunction(mainThread = true)
    public final void fire() {
        if(this.tileEntity.isRunning()) {
            this.tileEntity.getContraption().tryFiringShot();
        }
    }

    @LuaFunction(mainThread = true)
    public final boolean isRunning(){
        return this.tileEntity.isRunning();
    }

//    @LuaFunction
//    public final Object getPitchOffset(IArguments partialTicks) throws LuaException {
//        if(this.isRunning()) {
//            double value = partialTicks.optDouble(0).orElse(0.0);
//            return (double) this.tileEntity.getPitchOffset((float) value);
//        }
//        return false;
//    }
//
//    @LuaFunction
//    public final Object getYawOffset(IArguments partialTicks) throws LuaException {
//        if(this.isRunning()) {
//            double value = partialTicks.optDouble(0).orElse(0.0);
//            return (double) this.tileEntity.getYawOffset((float) value);
//        }
//        return false;
//    }

    @LuaFunction
    public final double getPitch() {
        return ((CannonMountBlockEntityAccessor)this.tileEntity).getCannonPitch();
    }

    @LuaFunction
    public final double getYaw() {
        return ((CannonMountBlockEntityAccessor)this.tileEntity).getCannonYaw();
    }

    @LuaFunction
    public final Object getMaxDepress() {
        return (double) this.tileEntity.getContraption().maximumDepression();
    }

    @LuaFunction
    public final Object getMaxElevate() {
        return (double) this.tileEntity.getContraption().maximumElevation();
    }
}

File: common/src/main/kotlin/io/github/xiewuzhiying/vs_addition/compats/computercraft/peripheral/CannonMountPeripheral.kt
package io.github.xiewuzhiying.vs_addition.compats.computercraft.peripheral

import dan200.computercraft.api.lua.LuaFunction
import dan200.computercraft.api.peripheral.IPeripheral
import io.github.xiewuzhiying.vs_addition.mixin.createbigcannons.CannonMountBlockEntityAccessor
import io.github.xiewuzhiying.vs_addition.mixinducks.createbigcannons.MountedAutocannonContraptionMixinDuck
import net.minecraft.server.level.ServerLevel
import rbasamoyai.createbigcannons.cannon_control.cannon_mount.CannonMountBlockEntity
import rbasamoyai.createbigcannons.cannon_control.contraption.AbstractMountedCannonContraption

open class CannonMountPeripheral(
    val peripheralType: String,
    val tileEntity: CannonMountBlockEntity
) : IPeripheral{
    override fun getType(): String {
        return peripheralType
    }

    override fun equals(iPeripheral: IPeripheral?): Boolean {
        return iPeripheral === this
    }

    override fun getTarget(): Any {
        return this.tileEntity
    }

    @LuaFunction(mainThread = true)
    fun assemble(): Any {
        if (!tileEntity.isRunning) {
            (tileEntity as CannonMountBlockEntityAccessor?)?.Assemble()
            return true
        }
        return false
    }

    @LuaFunction(mainThread = true)
    fun disassemble(): Any {
        if (tileEntity.isRunning) {
            tileEntity.disassemble()
            tileEntity.sendData()
            return true
        }
        return false
    }

    @LuaFunction(mainThread = true)
    fun fire() {
        if (this.tileEntity.contraption?.level() is ServerLevel) {
            (this.tileEntity.contraption?.contraption as? MountedAutocannonContraptionMixinDuck)?.setIsCalledByComputer()
            (this.tileEntity.contraption?.contraption as AbstractMountedCannonContraption).fireShot(tileEntity.contraption?.level() as ServerLevel, this.tileEntity.contraption)
        }
    }

    @LuaFunction(mainThread = true)
    fun isRunning(): Boolean {
        return tileEntity.isRunning
    }

    @LuaFunction
    fun getPitch(): Double {
        return (tileEntity as CannonMountBlockEntityAccessor).cannonPitch.toDouble()
    }

    @LuaFunction
    fun getYaw(): Double {
        return (tileEntity as CannonMountBlockEntityAccessor).cannonYaw.toDouble()
    }

    @LuaFunction
    fun getX(): Int {
        return tileEntity.controllerBlockPos.x
    }

    @LuaFunction
    fun getY(): Int {
        return tileEntity.controllerBlockPos.y
    }

    @LuaFunction
    fun getZ(): Int {
        return tileEntity.controllerBlockPos.z
    }

    @LuaFunction
    fun getMaxDepress(): Double? {
        return tileEntity.contraption?.maximumDepression()?.toDouble()
    }

    @LuaFunction
    fun getMaxElevate(): Double? {
        return tileEntity.contraption?.maximumElevation()?.toDouble()
    }
}